core/telemetry/src/lib.rs

```diff
@@ -43,66 +48,165 @@ pub type Telemetry = slog_scope::GlobalLoggerGuard;
+		out_syncs.push(out_sync);
+
+		let until_verbosity = *verbosity;
+		let filter = slog::Filter(
+			slog_json::Json::default(writer).fuse(),
+			move |rec| {
+				let tag = rec.tag().parse::<u8>()
+					.expect("`telemetry!` macro requires tag.");
+				tag <= until_verbosity
+			});
+
+		let filter = Box::new(filter) as Box<slog::Filter<_, _>>;
+		endpoint_drains.push(filter);
+	}
+
+	// Set up logging to all endpoints
+	let drain = slog_async::Async::new(Multiply::new(endpoint_drains).fuse());
+	let root = slog::Logger::root(drain.chan_size(CHANNEL_SIZE)
 		.overflow_strategy(slog_async::OverflowStrategy::DropAndReport)
 		.build().fuse(), o!()
 	);
-	let logger_guard = slog_scope::set_global_logger(log);
-
-	thread::spawn(move || {
-		loop {
-			trace!(target: "telemetry", "Connecting to Telemetry... {:?}", config.url);
-			let _ = ws::connect(config.url.as_str(), |out| Connection::new(out, &*out_sync, &config));
-
-			thread::sleep(time::Duration::from_millis(5000));
-		}
-	});
+	let logger_guard = slog_scope::set_global_logger(root);
+
+	// Spawn a thread for each endpoint
+	let on_connect = Arc::new(config.on_connect);
+	let mut iter = endpoints.iter();
+	while let Some((url, verbosity)) = iter.next() {
+		let url_ = url.clone();
+		let inner_url = Arc::new(url.to_owned());
+		let inner_verbosity = Arc::new(verbosity.to_owned());
+		let inner_on_connect = Arc::clone(&on_connect);
+
+		let out_sync = out_syncs.remove(0);
+		let out_sync = Arc::clone(&out_sync);
+
+		thread::spawn(move || {
+			loop {
+				let on_connect = Arc::clone(&inner_on_connect);
+				let out_sync = Arc::clone(&out_sync);
+				let url = Arc::clone(&inner_url);
+				let verbosity = Arc::clone(&inner_verbosity);
+
+				trace!(target: "telemetry",
+					"Connecting to Telemetry at {} with verbosity {}", Arc::clone(&url), Arc::clone(&verbosity));
+
+				let _ = ws::connect(url_.to_owned(),
+					move |out| {
+						Connection::new(out, Arc::clone(&out_sync), Arc::clone(&on_connect), Arc::clone(&url))
+					});
+
+				thread::sleep(time::Duration::from_millis(5000));
+			}
+		});
+	}
 
 	return logger_guard;
 }
 
-/// Exactly equivalent to `slog_scope::info`, provided as a convenience.
+/// Translates to `slog_scope::info`, but contains an additional verbosity
+/// parameter which the log record is tagged with. Additionally the verbosity
+/// parameter is added to the record as a key-value pair.
 #[macro_export]
 macro_rules! telemetry {
-	( $($t:tt)* ) => { $crate::with_logger(|l| $crate::slog::slog_info!(l, $($t)* )) }
+	 ( $a:expr; $b:expr; $( $t:tt )* ) => {
+		$crate::with_logger(|l| {
+			$crate::slog::slog_info!(l, #$a, $b; "verbosity" => stringify!($a), $($t)* )
+		})
+    }
 }
 
-struct Connection<'a> {
+struct Connection {
 	out: ws::Sender,
-	out_sync: &'a Mutex<Option<ws::Sender>>,
-	config: &'a TelemetryConfig,
+	out_sync: Arc<Mutex<Option<ws::Sender>>>,
+	on_connect: Arc<Box<Fn() + Send + Sync + 'static>>,
+	url: Arc<String>,
```

Comment:
Why is this an `Arc`?

Reply:
Haven't found a way around it. The issues is that I iterate through all endpoints and in that loop I spawn threads, in which I create closures. The url is needed within this closure and the compiler won't allow me to borrow it in the closure. I suspect that the issue is this looping thread-creation with nested closures.

Yeah it's a bit unfortunate, since the url is only needed for `trace!`.